### PR TITLE
[feat] PL mvp0: training

### PR DESCRIPTION
**Pull request by ytsheng**
* pytorch lighting stub mostly involving training
* Tests for lightning trainer included
* built on top of the mmf grad accumulation fix: https://github.com/facebookresearch/mmf/pull/747

- [X] MVP 0. Training: Goal - Train a model from scratch and reach similar accuracy as using mmf_trainer
   - [X] Setup the training pipeline: done
   - [X] Training on the right device: done
   - [X] Clip gradients: done
   - [X] Optimizer: done
   - [X] FP16 Support: done
   - [X] LR scheduler (incl. warmup etc): done
   - [X] testcase: train visual_bert on vqa from scratch fo 10 iterations, compare the value: done
- [x] tests included in this PR (tests are only done for pytorch lightning integration):
   - [X] Vanilla Training w/o grad accumulate, make sure loss for 5 iters are the same between mmf and pl 
      - [X] Optimizer working as intended as a part of this PR.
   - [X] `max_updates` and `max_epochs` calculation
   - [x] Training with grad accumulate
   - [x] Training with LR schedule achieves a different value compared to without LR schedule
   - [x] Training with LR schedule for PL is the same as training with LR schedule for `mmf_tranier`
   - [x] Training with gradient clipping make sure all grads are within the `grad_clipping` threshold
   - [x] Training with gradient clipping is the same as training with gradient clipping for `mmf_trainer`

### CODE REVIEW

**Comment by SeanNaren** [comment]
Apologies on the delay getting https://github.com/PyTorchLightning/pytorch-lightning/pull/4369 merged, a test was added so it should be merged today and added to the release!

**Comment by ytsheng** [comment]
Thanks so much @SeanNaren, will take a look!

**Comment by SeanNaren** [comment]
I gave the PR a read-over and looks clean, nice work! I saw some custom builders in a few places for optimizers/schedulers, have you guys thought of using the hydra instantiation methods: https://hydra.cc/docs/next/patterns/instantiate_objects/overview

I know hydra instantiations is not to everyones tastes :) Regardless integration looks great!

**Comment by ytsheng** [comment]
> I gave the PR a read-over and looks clean, nice work! I saw some custom builders in a few places for optimizers/schedulers, have you guys thought of using the hydra instantiation methods: https://hydra.cc/docs/next/patterns/instantiate_objects/overview
> 
> I know hydra instantiations is not to everyones tastes :) Regardless integration looks great!

Hydra is definitely in our pipeline of things to do! We have it planned for H1 of 2021. Stay tuned.

File: tests/trainers/test_training_loop.py [hunk 1]
```diff
@@ -113,9 +134,9 @@ def test_update_frequency_reporting(self):
         def _on_update_end(report, meter, should_log):
             # the losses here should be the sum of two losses in
             # iteration 0 and iteration 1 (both constitute update 0).
-            # Here iter 1 loss: 0.2684, iter 2 loss: 2.4167
+            # Here iter 1 loss: 0.2599, iter 2 loss: 4.2090
             loss = report.losses["loss"].detach().cpu().item()
-            self.assertAlmostEqual(loss, 2.6852, 4)
+            self.assertAlmostEqual(loss, 4.4688, 4)
```
Comment (ytsheng) [inline tests/trainers/test_training_loop.py:139]: this was necessary because I changed the loss calculation to force gradients to be big.

File: tests/test_utils.py [hunk 2]
```diff
@@ -120,10 +119,42 @@ def __init__(self, size):
...
+
+            optimizer = build_optimizer(self, self.config)
+            if self.config.training.lr_scheduler:
+                lr_scheduler = build_scheduler(optimizer, self.config)
+                return {
+                    "optimizer": optimizer,
+                    "lr_scheduler": {"scheduler": lr_scheduler, "interval": "step"},
+                }
+            else:
+                return optimizer
```
Comment (apsdehal) [inline tests/test_utils.py:157]: Might be better to refactor this to a function as it is used in BaseModel as well.

File: requirements.txt [hunk 3]
```diff
@@ -15,3 +15,4 @@ omegaconf==2.0.1rc4
 lmdb==0.98
 termcolor==1.1.0
 iopath==0.1.3
+pytorch_lightning>=1.1.6
```
Comment (apsdehal) [inline requirements.txt:18]: We should fix the version instead of `>=`.

File: mmf/trainers/lightning_trainer.py [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+            max_steps=self._max_updates,
+            max_epochs=self._max_epochs,
+            gradient_clip_val=self._gradient_clip_val,
+            num_sanity_val_steps=0,
+            automatic_optimization=True,
+            checkpoint_callback=False,
+            accumulate_grad_batches=self.training_config.update_frequency,
+            val_check_interval=self.training_config.evaluation_interval,
+            log_every_n_steps=self.training_config.log_interval,
+        )
```
Comment (apsdehal) [inline mmf/trainers/lightning_trainer.py:51]: Here is the problem I was talking about. Now, users need to learn extra stuff such as what do lightning parameters correspond to in MMF which is unnecessary. We are losing one of the benefits of moving to lightning as these parameters would be better documented in Lightning. 

Also, if something else is added here tomorrow, we will need to again update our training creating an unnecessary dependency loop.

File: mmf/trainers/lightning_trainer.py [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+        self._calculate_gradient_clip_val()
+        self._load_trainer()
+
+    def _load_trainer(self):
+        self.trainer = Trainer(
+            logger=False,
+            gpus=self._gpus,
+            num_nodes=self._num_nodes,
+            callbacks=self._callbacks,
+            precision=16 if self.training_config.fp16 else 32,
```
Comment (apsdehal) [inline mmf/trainers/lightning_trainer.py:39]: Please refactor these to constants and then use them.

File: mmf/trainers/lightning_trainer.py [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+
+    def configure_seed(self):
+        default_deterministic = False
+        seed = self.config.training.seed
+
+        self._deterministic = default_deterministic
+        if seed is not None:
+            self._deterministic = True
+
+        self._benchmark = False
```
Comment (apsdehal) [inline mmf/trainers/lightning_trainer.py:78]: User doesn't even have the ability to set them?

File: mmf/trainers/lightning_trainer.py [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+            callbacks=self._callbacks,
+            precision=16 if self.training_config.fp16 else 32,
+            deterministic=self._deterministic,
+            benchmark=self._benchmark,
+            max_steps=self._max_updates,
+            max_epochs=self._max_epochs,
+            gradient_clip_val=self._gradient_clip_val,
+            num_sanity_val_steps=0,
+            automatic_optimization=True,
+            checkpoint_callback=False,
```
Comment (apsdehal) [inline mmf/trainers/lightning_trainer.py:47]: This is more of the same problem, user can't even override these choices now.

File: mmf/trainers/lightning_trainer.py [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+
+    def load_optimizer(self):
+        logger.info("Loading optimizer")
+        if self.trainer:
+            pass
+            # TODO: sash, this is a problem, needs fix
+            # at this point the model's configure_optimizer
+            # function isnt called yet, so this will break.
+            # self.optimizer = self.trainer.optimizers[0]
+            # we need an assigned optimizer for checkpoint
```
Comment (apsdehal) [inline mmf/trainers/lightning_trainer.py:106]: You can also refactor the code to expect the trainer to have optimzer property.

File: tests/lightning/test_lightning_trainer.py [hunk 9]
```diff
@@ -0,0 +1,346 @@
...
+    def _prepare_trainer(self, trainer):
+        trainer.configure_device()
+        trainer._calculate_max_updates()
+        trainer._calculate_gradient_clip_val()
+        trainer._load_trainer()
+
+    def test_same_loss_computation(self):
+        # check to see the same losses between the two trainers
+        # under the same conditions
+        self._test_same_loss_computation = True
```
Comment (apsdehal) [inline tests/lightning/test_lightning_trainer.py:301]: Is there a better approach to do this, this adds dynamic dependency between the tests and make them dependent certain conditions. Won't it be better to add new callback just for the sake of testing to the trainer?

**Review by apsdehal — COMMENTED** [review]
Super! This is going great. This needs to be imported internally and then TARGETS needed to be added in fbcode before landing. I have left some general comments on design.

File: mmf/trainers/lightning_trainer.py (hunk 8 above)
Comment (ytsheng) [reply mmf/trainers/lightning_trainer.py:106]: Addressed in the next PR.

File: mmf/trainers/lightning_trainer.py (hunk 6 above)
Comment (ytsheng) [reply mmf/trainers/lightning_trainer.py:78]: This behavior is consistent with mmf

File: mmf/models/base_model.py [hunk 10]
```diff
@@ -163,18 +183,81 @@ def forward(self, sample_list, *args, **kwargs):
...
+            current iteration
+
+        Returns:
+            Dict: Dict containing loss.
+        """
+        batch = self._ensure_sample_list(batch)
+        output = self(batch)
+        loss_dict = output["losses"]
+        output["loss"] = sum([loss.mean() for loss in loss_dict.values()])
+        return output
```
[the remark this answers is not in the document]
Comment (ytsheng) [reply mmf/models/base_model.py:202]: addressed in the next PR. :)

File: mmf/datasets/base_dataset.py [hunk 11]
```diff
@@ -65,7 +65,7 @@ def init_processors(self):
 
     def prepare_batch(self, batch):
         """
-        Can be possibly overridden in your child class
+        Can be possibly overridden in your child class. Deprecated w Pytorch Lightning
```
[the remark this answers is not in the document]
Comment (vedanuj) [reply mmf/datasets/base_dataset.py:68]:
```suggestion
        Can be possibly overridden in your child class. Deprecated w Lightning trainer
```

File: mmf/trainers/lightning_trainer.py (hunk 4 above)
Comment (vedanuj) [reply mmf/trainers/lightning_trainer.py:51]: Agree. These should be passed as parameters to the lightning trainer config directly.

File: mmf/trainers/lightning_trainer.py (hunk 6 above)
Comment (apsdehal) [reply mmf/trainers/lightning_trainer.py:78]: What I meant was since lightning provides this functionality, we should allow user to use it.

File: mmf/configs/defaults.yaml [hunk 12]
```diff
@@ -4,6 +4,7 @@ config_version: 1.0
 # Configuration for training
 training:
```
[the remark this answers is not in the document]
Comment (ytsheng) [reply mmf/configs/defaults.yaml:5]: `type, params` is deprecated in Hydra though, do we still want to use `type, params`?  Alternatively, to isolate lightning params, I can create another key for lightning, which will be turned on when the `training.trainer = lightning`

File: mmf/datasets/lightning_datamodule.py [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+
+        self.train_loader = MultiDatasetLoader("train")
+        self.val_loader = MultiDatasetLoader("val")
+        self.test_loader = MultiDatasetLoader("test")
+
+        self.train_loader.load(self.config)
+        self.val_loader.load(self.config)
+        self.test_loader.load(self.config)
+
+    def prepare_data(self):
```
Comment (apsdehal) [inline mmf/datasets/lightning_datamodule.py:24]: Do they need to be here if they are not going to be used?

File: mmf/datasets/lightning_datamodule.py [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+        pass
+
+    def setup(self, stage: Optional[str] = None):
```
Comment (apsdehal) [inline mmf/datasets/lightning_datamodule.py:27]: Same!

File: mmf/trainers/lightning_core/loop_callback.py [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+
+logger = logging.getLogger(__name__)
+
+
+class LightningLoopCallback(Callback):
+    def __init__(self, lightning_trainer):
+        super().__init__()
+        self.lightning_trainer = lightning_trainer
+
+    def on_init_start(self, trainer):
```
Comment (apsdehal) [inline mmf/trainers/lightning_core/loop_callback.py:17]: Same. is this needed?

File: mmf/trainers/lightning_trainer.py [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+        super().load()
+        self.trainer_config = self.config.trainer.params
+        self._calculate_max_updates()
+        self._load_trainer()
+
+    def _load_trainer(self):
+        lightning_params = self.trainer_config
+        self.trainer = Trainer(
+            logger=False,
+            gpus=lightning_params.gpus,
```
Comment (apsdehal) [inline mmf/trainers/lightning_trainer.py:36]: You should pass ligthning_params as `**lightning_params` so that if something comes in tomorrow, it can be easily overridden in child configs.

**Review by apsdehal — COMMENTED** [review]
Looking good! Almost there.

File: tests/lightning/__init__.py [hunk 17]
```diff
@@ -0,0 +1 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
```
Comment (vedanuj) [inline tests/lightning/__init__.py:1]: The tests should ideally under `tests/trainers` directory to maintain the current test directory structure. You can keep the `lightning' subdirectory.

**Review by vedanuj — COMMENTED** [review]
Looks good to me overall. Some comments to address before landing.

**Review by vedanuj — APPROVED** [review]
Looks good to me! Thanks for addressing all the comments.
